[RDY] :move without closing folds

hardenedapple opened the pull request:
### Problem description
The fix [here](https://groups.google.com/forum/#!msg/vim_dev/5SD5E_hyGPk/-ojYGyeNbwYJ)
for the problem with `:move` corrupting folds doesn't account for manual folds,
and has the unfortunate side effect of closing any open folds.

I've made an alternate fix, that doesn't have either of these problems.

@chrisbra -- I believe this fix would work in vim, but I haven't checked yet.

### Underlying problem
The cause of the underlying bug was that `do_move()` calls `mark_adjust()` to move marks around.
`mark_adjust()` then calls `foldMarkAdjust()` to do the same with folds.
`foldMarkAdjust()` assumes that it is being called for a simple shift occuring
from either deletion or insertion of lines. It hence moves folds without
checking if it causes overlap or leaves folds out of order in the growarray of
folds.
Once folds are out of order, `findFold()` isn't guaranteed to find the correct
fold, and hence the `foldUpdate()` call is compromised too.
Similarly, overlapping folds breaks some assumptions in
`foldUpdateIEMSRecurse()`, but I didn't look into what problems come from what
assumptions.

`do_move()` is the only place that `mark_adjust()` is called for something
other than a simple addition or deletion of lines, and hence the only place
that this problem is observed. (`do_filter()` actually also calls
`mark_adjust()` in a way that I suspect can cause bugs -- search for 
`"if (do_in)"` and see where marks are moved _then_ deleted -- but I haven't
checked/prooved there's a bug there yet so I haven't fixed it, and the fix
would simply be to swap the order of calls to `mark_adjust()` around there)

### My changes
I'm keeping the current behaviour of `foldMarkAdjust()`, as I don't believe
there's a clear way to handle all cases of moving folds so they overlap (e.g.
what should a generalised version of `foldMarkAdjust()` do if it is told to
move a range of folds down such that the top of the next fold is now within a
fold contained in the range that has been moved).

Instead, in `do_move()` I'm moving marks around in the same way they were
before but without moving the folds (by introducing a new function called
`mark_adjust_nofold()`, and I'm directly moving the folds with a new function
called `foldSwapRange()`.

### WIP because
1) I plan on checking a suspected bug in `do_filter()`.
2) I wanted to ask others if this seems like a reasonable fix in the meantime
(to hear any faults others might see).

hardenedapple committed:
Robustly handle folds during a :move command

In order to re-order marks according to the :move command, do_move()
uses mark_adjust() in a non-standard manner. The non-standard action is
that it moves some marks *past* other marks. This doesn't matter for
marks, but mark_adjust() calls foldMarkAdjust() which simply changes
fold starts and lengths and doesn't have enough information to know that
other folds have to be checked and reordered.

The array of folds for each window are assumed to be in order of
increasing line number, and if this gets broken some folds can get
"lost".

There has been a previous patch to avoid this problem by deleting and
recalculating all folds in the window, but this comes at the cost of
closing all folds when executing :move, and doesn't cover the case of
manual folds.
This patch adds a new function foldMoveRange() specifically for the
:move command that handles reordering folds as well as simply moving
them. Additionally, we allow calling mark_adjust_nofold() that does the
same as mark_adjust() but doesn't affect any fold array.

Calling mark_adjust_nofold() should be done in the same manner as
calling mark_adjust(), but according changes to the fold arrays must be
done seperately by the calling function.

vim-patch:8.0.0457
vim-patch:8.0.0459
vim-patch:8.0.0461
vim-patch:8.0.0465

justinmk commented on src/nvim/mark.c:
```diff
@@ -887,6 +887,18 @@ void ex_changes(exarg_T *eap)
+}
+
+static void mark_adjust_internal(linenr_T line1, linenr_T line2, long amount,
+    long amount_after, bool move_folds)
```
perhaps call it `adjust_folds` to keep the jargon consistent.

hardenedapple replied:
Sorry: I'm not sure what you mean -- which function to be called `adjust_folds()`?

hardenedapple replied:
Ah -- I get it now -- you mean the argument `move_folds` should be `adjust_folds`.

justinmk commented on src/nvim/ex_cmds.c:
```diff
@@ -801,21 +793,24 @@ int do_move(linenr_T line1, linenr_T line2, linenr_T dest)
-  mark_adjust(last_line - num_lines + 1, last_line,
+  mark_adjust_nofold(last_line - num_lines + 1, last_line,
       -(last_line - dest - extra), 0L);
-  changed_lines(last_line - num_lines + 1, 0, last_line + 1, -extra);
```
why doesn't `changed_lines` need to be called after `mark_adjust_nofold`?

hardenedapple replied:
`changed_lines()` is needed after `mark_adjust_nofold()` -- it's called at the end of the `do_move()` function.

This is similar to what happened originally (before the previous fix was done).
Before, `mark_adjust()` was used to move marks around bit-by-bit, to get them in the correct place (because it's not a simple move everything down/up), then the extra lines from the initial copy were deleted, and `changed_lines()` was called.
Now I've replaced `mark_adjust()` with `mark_adjust_nofolds()`, and dealt with the folds independently.

This call to `changed_lines()` is a later addition that fixed part of the problem [see here](https://groups.google.com/forum/#!msg/vim_dev/5SD5E_hyGPk/-ojYGyeNbwYJ).

The single call to `foldSwapRange()` puts the folds in the correct place (assuming no bugs ...) where it is in this paragraph doesn't actually matter, as nothing else in the block modifies or uses the fold positions. I put it in the existing `if/else` because I needed an `if/else` condition on the boolean that was already there.
It may not be clear in the code, but the gist of what I've done is to completely separate the updating of the marks and folds.
If you imagine three layers -- the text, the marks, and the folds -- then `mark_adjust()` moves both the marks and folds down/up at once, while `mark_adjust_nofold()` just moves marks and `foldSwapRange()` just swaps folds.
As long as each layer lines up correctly in the end the order they're called doesn't matter, and you can call them as many times as you want.

justinmk commented on src/nvim/mark.c:
```diff
@@ -887,6 +887,18 @@ void ex_changes(exarg_T *eap)
+  mark_adjust_internal(line1, line2, amount, amount_after, true);
+}
+
+void mark_adjust_nofold(linenr_T line1, linenr_T line2, long amount,
```
The doc for `mark_adjust` mentions:

> Must be called before changed_*(), appended_lines() or deleted_lines().

and it's a common pattern throughout the code. So the fact that `mark_adjust_nofold` _doesn't_ work like that could be surprising/confusing. To start with, the doc for `mark_adjust_nofold` should emphasize this situation, and I guess also mention the relationship to `foldSwapRange()`.

hardenedapple replied:
`mark_adjust_nofold()` does work in the same way as `mark_adjust()` (apart from needing updates of the folds separately).

The way I read that bit of the documentation is:
 when you're calling `changed_*() ...`, make sure to call `mark_adjust()` first.
Not 
make sure to call one of `changed_*() ...` after each call to `mark_adjust()`.

Calling `mark_adjust()` more than once before calling one of those mentioned functions is also done in `do_filter()` -- the same place that I think there's a bug, but not because of that.

Agreed on the documentation about the relationship with folds, but just by way of explanation: there's no special relationship between `mark_adjust_nofold()` and `foldSwapRange()`.
`foldSwapRange()` was just a function to do the specific moving around of folds that I needed in this case.

hardenedapple commented:
It turns out that the problem I thought would be in `do_filter()` isn't really a problem.

The lines i was concerned about can indeed put folds out of order, but because of the way we then call `del_lines()` with the current cursor before the range that has changed, this incorrect order is then fixed.
This happens because 1) searching for the fold beginning before the range that has been corrupted with `foldFind()` still gives you the correct place, as all folds to the right of the index found will be greater than the line searched for, whether out of order or not. 2) when `foldMarkAdjustRecurse()` is eventually called with a non-zero `amount_after`, it iterates over all folds in the window array after the one it's found (which will be correct because of point 1). It adjusts each one according to the values of that fold structure, and hence the order doesn't matter.

Also, the change I thought would have fixed the problem were it there, doesn't actually change anything to do with folds, as the first `if` statement in `foldMarkAdjust()` means one of the `mark_adjust()` calls in `do_filter()` doesn't change the folds at all.

Hence I'm not changing anything there.

chrisbra commented:
What exactly does 87b8d2 fix and why is this necessary?

hardenedapple commented:
The most obvious case 87b8d2 fixes is when you're moving lines within a given fold.
i.e. with the contents

```
a
    a
    a
    a
    a
```

If you set folds with
```vimL
set foldmethod=indent
set foldmethod=manual
```

then run the command `3m4`, before that commit, this would truncate the fold to start and end on line 2, now it doesn't change the fold.

I've actually just noticed another problem while writing this comment -- the same case but with the command `3m5`

hardenedapple commented:
Just for reference, upstream have included this patch in [a](https://github.com/vim/vim/releases/tag/v8.0.0465) [few](https://github.com/vim/vim/releases/tag/v8.0.0461) [recent](https://github.com/vim/vim/releases/tag/v8.0.0459) [patches](https://github.com/vim/vim/releases/tag/v8.0.0457). So I changed the commit message to add the `vim-patch` tag.

jamessan commented:
> So I changed the commit message to add the vim-patch tag.

To ensure proper consumption by other tools, can you use a separate `vim-patch:...` line for each patch?

hardenedapple commented:
As in split the patch into multiple ones so that there is one commit per `vim-patch:...` line?

I can do that but it'll be a little ugly, as the vim patches do some extra stuff then remove it, and one is only there because I had an off-by-one error in the PR when it was taken that had to be fixed later.

To ask an opinion on the matter, which do other people think would be least confusing: `vim-patch` tags matching the equivalent vim tags but with the commits not the best separated, or `vim-patch` tags that don't directly correspond to the vim commits but instead try to separate logical steps in the commit?

I'll do the change once I've figured out why I'm getting an out of memory error in the 32 bit travis CI build.

jamessan commented:
> As in split the patch into multiple ones so that there is one commit per vim-patch:... line?

No, I just mean change the single line you have in the current commit message into a `vim-patch:` line per patch.

```
vim-patch:8.0.0457
vim-patch:8.0.0459
vim-patch:8.0.0461
vim-patch:8.0.0465
```

hardenedapple committed:
Fix wrap-around in 32 bit

hardenedapple committed:
vim-patch:8.0.0407 : filtering folds with marker method not tested

Problem:    Filtering folds with marker method not tested.
Solution:   Also set 'foldmethod' to "marker".

justinmk commented on src/nvim/fold.c:
```diff
@@ -2591,11 +2594,159 @@ static void foldRemove(garray_T *gap, linenr_T top, linenr_T bot)
+
+#define fold_end(fp) ((fp)->fd_top + (fp)->fd_len - 1)
+#define valid_fold(fp, gap) ((fp) < ((fold_T *)(gap)->ga_data + (gap)->ga_len))
+#define fold_index(fp, gap) ((size_t)(fp - ((fold_T *)(gap)->ga_data)))
```
these should be uppercase
